sklearn/utils/validation.py

```diff
@@ -1588,3 +1588,31 @@ def _check_fit_params(X, fit_params, indices=None):
             )
 
     return fit_params_validated
+
+
+def _get_feature_names(X):
+    """Get feature names from X.
+
+    Support for other array containers should place its implementation here.
+
+    Parameters
+    ----------
+    X : dataframe of shape (n_samples, n_features)
+        Array container to extract feature names.
+
+        - pandas DataFrame : The columns will be considered to be feature
+          names.
+        - All other array containers will return `None`.
+
+    Returns
+    -------
+    names: Sequence of str or None
+        Column names of `X`. Unrecognized array containers will return `None`.
+    """
+    if hasattr(X, "iloc"):
```

Comment:
why do we check if it's a `DataFrame` AND that it has `iloc`?

Reply:
Currently this checks for `iloc` first and then makes sure it is a dataframe, which is more strict. It is a little different from our normal assumption that "iloc" -> always a DataFrame. I do not have a overly strong opinion either way, so I can revert to just checking for `iloc`.

Reply:
I just find if odd to check for both, I'm agnostic on which one to use, ducktype or instance type.

Reply:
I like ducktyping ;)